apps/cpp/avifgainmaputil.cc

```diff
@@ -0,0 +1,386 @@
+    }
+    result = avifDecoderParse(decoder.get());
+    if (result != AVIF_RESULT_OK) {
+      std::cerr << "Failed to parse image: " << avifResultToString(result)
+                << " (" << decoder->diag.error << ")\n";
+      return result;
+    }
+    if (!decoder->gainMapPresent) {
+      std::cerr << "Input image " << input << " does not contain a gain map\n";
+      return AVIF_RESULT_INVALID_ARGUMENT;
+    }
+
+    const avifGainMapMetadata& metadata = decoder->image->gainMap.metadata;
+    const int w = 20;
+    std::cout << std::left << std::setw(w) << "Base headroom: "
+              << FormatFraction(metadata.baseHdrHeadroomN,
+                                metadata.baseHdrHeadroomD)
+              << "\n";
+    std::cout << std::left << std::setw(w) << "Alternate headroom: "
+              << FormatFraction(metadata.alternateHdrHeadroomN,
+                                metadata.alternateHdrHeadroomD)
+              << "\n";
+    std::cout << std::left << std::setw(w) << "Gain Map Min: "
+              << FormatFractions(metadata.gainMapMinN, metadata.gainMapMinD)
+              << "\n";
+    std::cout << std::left << std::setw(w) << "Gain Map Max: "
+              << FormatFractions(metadata.gainMapMaxN, metadata.gainMapMaxD)
+              << "\n";
+    std::cout << std::left << std::setw(w) << "Base Offset: "
+              << FormatFractions(metadata.baseOffsetN, metadata.baseOffsetD)
+              << "\n";
+    std::cout << std::left << std::setw(w) << "Alternate Offset: "
+              << FormatFractions(metadata.alternateOffsetN,
+                                 metadata.alternateOffsetD)
+              << "\n";
+    std::cout << std::left << std::setw(w) << "Gain Map Gamma: "
+              << FormatFractions(metadata.gainMapGammaN, metadata.gainMapGammaD)
+              << "\n";
+    std::cout << std::left << std::setw(w) << "Backward Direction: "
+              << (metadata.backwardDirection ? "True" : "False") << "\n";
+
+    return AVIF_RESULT_OK;
+  }
+};
+
+void PrintUsage(const std::vector<std::unique_ptr<ProgramCommand>>& commands) {
+  std::cout << "\nExperimental tool to manipulate avif images with HDR gain "
+               "maps.\n\n";
+  std::cout << "Usage: avifgainmaputil <command> [options] [arguments...]\n\n";
+  std::cout << "Available commands:\n";
+  int longest_command_size = 0;
+  for (const std::unique_ptr<ProgramCommand>& command : commands) {
+    longest_command_size =
+        std::max(longest_command_size, (int)command->GetName().size());
+  }
+  for (const std::unique_ptr<ProgramCommand>& command : commands) {
+    std::cout << "  " << std::left << std::setw(longest_command_size)
+              << command->GetName() << "  " << command->GetDescription()
+              << "\n";
+  }
+  std::cout << "\n";
+  avifPrintVersions();
+}
+
+}  // namespace
+
+MAIN() {
+  std::vector<std::unique_ptr<ProgramCommand>> commands;
+  commands.emplace_back(std::make_unique<HelpCommand>());
```

Comment:
maybe an initializer list is easier to read?

Reply:
I tried that first but std::vector *copies* the content of the initializer list which means all the commands must be copyable.